File: ui/src/lib/googleWorkspace.ts
/**
 * Hand-written API wrapper for the Google connection and workspace discovery
 * endpoints (`api/routes/integrations_google.py`).
 *
 * Hand-written rather than generated: `npm run generate-client` reads the live
 * backend's OpenAPI document, and these routes ship in the same change as the
 * page below them — there is no running backend to read them from yet. Re-run
 * the generator once the backend is up and this module can be replaced by
 * `@/client/sdk.gen` imports; the shapes here mirror `api/schemas/
 * google_workspace.py` and `api/schemas/google_sheets.py` field for field.
 *
 * Same mechanics as `lib/platformAdminApi.ts`: requests go through the shared
 * generated `client`, inheriting its base URL and the auth interceptor
 * registered once auth is loaded. Callers must guard on `authLoading`/`user`
 * (see ui/AGENTS.md).
 *
 * Errors come back as `GoogleRequestError`, not bare strings, because the page
 * has to branch on *why* a call failed: a missing scope offers "grant this
 * permission", a dead grant offers "reconnect", and a 503 means the deployment
 * has no OAuth client at all. `message` is always a plain string — the
 * backend's `detail` is a dict here and must never reach JSX.
 */

import { client } from "@/client/client.gen";
import { detailFromError } from "@/lib/apiError";

// ---------------------------------------------------------------------------
// Connection
// ---------------------------------------------------------------------------

/** Browser-safe deployment config: no client secret has a field here. */
export interface GoogleOAuthPublicConfig {
    enabled: boolean;
    client_id?: string | null;
    picker_api_key?: string | null;
    picker_app_id?: string | null;
    scopes: string[];
}

export interface GoogleConnectionStatus {
    connected: boolean;
    /** 'connected' or 'revoked'; null when the organization never connected. */
    status?: string | null;
    account_email?: string | null;
    /** Access-token expiry, not an expiry of the connection — the backend refreshes it. */
    expires_at?: string | null;
    scopes: string[];
    requires_reconsent: boolean;
    revoked_reason?: string | null;
    oauth_config: GoogleOAuthPublicConfig;
}

export interface GoogleAuthorizeResponse {
    authorization_url: string;
    redirect_uri: string;
    expires_in_seconds: number;
}

export interface GoogleDisconnectResponse {
    disconnected: boolean;
}

// ---------------------------------------------------------------------------
// Resource summaries
// ---------------------------------------------------------------------------

export interface SpreadsheetSummary {
    id: string;
    name: string;
    web_url?: string | null;
    modified_time?: string | null;
}

export interface SheetTabSummary {
    sheet_id?: number | null;
    title: string;
    index?: number | null;
    row_count?: number | null;
    column_count?: number | null;
}

export interface CalendarSummary {
    id: string;
    summary?: string | null;
    description?: string | null;
    primary: boolean;
    /** 'owner' | 'writer' | 'reader' | 'freeBusyReader' — below 'writer' cannot host bookings. */
    access_role?: string | null;
    color_id?: string | null;
    time_zone?: string | null;
}

export interface DriveItemSummary {
    id: string;
    name: string;
    mime_type?: string | null;
    is_folder: boolean;
    web_url?: string | null;
    modified_time?: string | null;
}

export interface GmailLabelSummary {
    id: string;
    name: string;
    type?: string | null;
}

export interface GmailMessageSummary {
    id: string;
    thread_id?: string | null;
    subject?: string | null;
    /** The From header verbatim. Named `from_address` because `from` is a Python keyword. */
    from_address?: string | null;
    /** Kept as text: senders emit malformed dates and a parse failure must not drop the row. */
    date?: string | null;
    snippet?: string | null;
}

export interface DocsSummary {
    id: string;
    name: string;
    web_url?: string | null;
    modified_time?: string | null;
}

// ---------------------------------------------------------------------------
// Workspace overview
// ---------------------------------------------------------------------------

/**
 * Why a section looks the way it does. Read these before the items: an empty
 * section with `scope_missing` is a permission to ask for, one with `error` is
 * Google failing, and one with neither is an account that has nothing there.
 */
export interface WorkspaceSectionFlags {
    available: boolean;
    scope_missing: boolean;
    missing_scopes: string[];
    error?: string | null;
    /**
     * Set when the section worked but the grant narrows what it can see, so a
     * short or empty list is expected rather than wrong. Distinct from
     * `error`: nothing failed, and distinct from `scope_missing`: the scopes
     * this section declares WERE granted.
     */
    notice?: string | null;
}

export interface WorkspaceSectionResponse<ItemT> extends WorkspaceSectionFlags {
    items: ItemT[];
}

export interface WorkspaceDriveSection extends WorkspaceSectionFlags {
    recent_files: DriveItemSummary[];
    root_folders: DriveItemSummary[];
}

export interface WorkspaceGmailSection extends WorkspaceSectionFlags {
    labels: GmailLabelSummary[];
    recent: GmailMessageSummary[];
}

export interface WorkspaceOverview {
    connected: boolean;
    google_email?: string | null;
    status?: string | null;
    requires_reconsent: boolean;
    /** What this organization consented to. */
    granted_scopes: string[];
    /** What the deployment asks for at all — a scope absent here cannot be re-consented into. */
    requested_scopes: string[];
    sheets: WorkspaceSectionResponse<SpreadsheetSummary>;
    calendars: WorkspaceSectionResponse<CalendarSummary>;
    drive: WorkspaceDriveSection;
    gmail: WorkspaceGmailSection;
    docs: WorkspaceSectionResponse<DocsSummary>;
}

// ---------------------------------------------------------------------------
// Scope registry
// ---------------------------------------------------------------------------

export interface GoogleScopeInfo {
    scope: string;
    /** 'non_sensitive' | 'sensitive' | 'restricted' | 'unknown'. */
    sensitivity: string;
    requested: boolean;
    granted: boolean;
}

export interface GoogleRequiredScopes {
    /** Scopes per family — 'sheets', 'calendar', 'drive', 'gmail', 'docs', plus 'all'. */
    families: Record<string, string[]>;
    requested_scopes: string[];
    deployment_sensitivity: string;
    scope_details: GoogleScopeInfo[];
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

/** A failed Google call, with the fields the UI branches on. */
export class GoogleRequestError extends Error {
    readonly status: number | undefined;
    /** Backend error code, e.g. 'insufficient_scope', 'not_configured'. */
    readonly code: string | undefined;
    /** Scope to add to the consent, when the grant was too narrow. */
    readonly missingScope: string | undefined;
    /** True when only a new trip through the consent screen can fix it. */
    readonly requiresReconsent: boolean;

    constructor(
        message: string,
        fields: {
            status?: number;
            code?: string;
            missingScope?: string;
            requiresReconsent?: boolean;
        } = {},
    ) {
        super(message);
        this.name = "GoogleRequestError";
        this.status = fields.status;
        this.code = fields.code;
        this.missingScope = fields.missingScope;
        this.requiresReconsent = Boolean(fields.requiresReconsent);
    }
}

/** Message to show for any thrown value — never an object, never a raw `detail`. */
export function googleErrorMessage(err: unknown, fallback = "Google request failed"): string {
    if (err instanceof Error && err.message) return err.message;
    return detailFromError(err, fallback);
}

function toGoogleError(
    err: unknown,
    status: number | undefined,
    fallback: string,
): GoogleRequestError {
    // These routes answer with `detail` as a dict; `detailFromError` only knows
    // the string and 422-array shapes, so the dict is unpacked here and the
    // helper stays the fallback for everything else.
    const detail = (err as { detail?: unknown } | null | undefined)?.detail;
    const fields =
        detail && typeof detail === "object" && !Array.isArray(detail)
            ? (detail as Record<string, unknown>)
            : null;

    const message =
        fields && typeof fields.message === "string" && fields.message
            ? fields.message
            : detailFromError(err, fallback);

    return new GoogleRequestError(message, {
        status,
        code: typeof fields?.code === "string" ? fields.code : undefined,
        missingScope:
            typeof fields?.missing_scope === "string" ? fields.missing_scope : undefined,
        requiresReconsent: fields?.requires_reconsent === true,
    });
}

// ---------------------------------------------------------------------------
// Transport
// ---------------------------------------------------------------------------

const BASE = "/api/v1/integrations/google";

async function request<T>(
    method: "get" | "delete",
    url: string,
    fallback: string,
    query?: Record<string, unknown>,
): Promise<T> {
    const res = await client[method]({ url, query });
    if (res.error !== undefined) {
        throw toGoogleError(res.error, res.response?.status, fallback);
    }
    return res.data as T;
}

// ---------------------------------------------------------------------------
// Connection
// ---------------------------------------------------------------------------

export async function fetchGoogleStatus(): Promise<GoogleConnectionStatus> {
    return request<GoogleConnectionStatus>(
        "get",
        `${BASE}/status`,
        "Failed to read the Google connection status",
    );
}

/**
 * Public: the scope registry this deployment was configured with. Fetched
 * before any connection exists so the page never hardcodes a scope list of its
 * own and drifts from the server's.
 */
export async function fetchGoogleRequiredScopes(): Promise<GoogleRequiredScopes> {
    return request<GoogleRequiredScopes>(
        "get",
        `${BASE}/oauth/required-scopes`,
        "Failed to read the Google permission list",
    );
}

/**
 * Mint a consent URL for the caller's organization. The scopes and the signed,
 * organization-bound state are chosen server-side — the browser only follows
 * the URL it gets back.
 */
export async function startGoogleAuthorization(): Promise<GoogleAuthorizeResponse> {
    return request<GoogleAuthorizeResponse>(
        "get",
        `${BASE}/authorize`,
        "Failed to start the Google connection",
    );
}

/** Idempotent: an organization with nothing connected gets `disconnected: false`. */
export async function disconnectGoogle(): Promise<GoogleDisconnectResponse> {
    return request<GoogleDisconnectResponse>(
        "delete",
        `${BASE}/connection`,
        "Failed to disconnect the Google account",
    );
}

// ---------------------------------------------------------------------------
// Discovery
// ---------------------------------------------------------------------------

export async function fetchWorkspaceOverview(): Promise<WorkspaceOverview> {
    return request<WorkspaceOverview>(
        "get",
        `${BASE}/workspace/overview`,
        "Failed to load your Google workspace",
    );
}

export async function listSpreadsheets(): Promise<SpreadsheetSummary[]> {
    return request<SpreadsheetSummary[]>(
        "get",
        `${BASE}/tools/sheets`,
        "Failed to list spreadsheets",
    );
}

export async function listSheetTabs(spreadsheetId: string): Promise<SheetTabSummary[]> {
    return request<SheetTabSummary[]>(
        "get",
        `${BASE}/tools/sheets/${encodeURIComponent(spreadsheetId)}/tabs`,
        "Failed to list the spreadsheet tabs",
    );
}

/**
 * First cells of a spreadsheet. Rows are ragged — Google trims trailing empty
 * cells — so pad against the longest row rather than assuming a rectangle.
 */
export async function previewSpreadsheet(
    spreadsheetId: string,
    range?: string,
): Promise<unknown[][]> {
    return request<unknown[][]>(
        "get",
        `${BASE}/tools/sheets/${encodeURIComponent(spreadsheetId)}/preview`,
        "Failed to preview the spreadsheet",
        range ? { range } : undefined,
    );
}

export async function listCalendars(): Promise<CalendarSummary[]> {
    return request<CalendarSummary[]>(
        "get",
        `${BASE}/tools/calendar/calendars`,
        "Failed to list calendars",
    );
}

export async function listDriveFolders(parentId?: string): Promise<DriveItemSummary[]> {
    return request<DriveItemSummary[]>(
        "get",
        `${BASE}/tools/drive/folders`,
        "Failed to list Drive folders",
        parentId ? { parent_id: parentId } : undefined,
    );
}

export async function listDriveFiles(params?: {
    folderId?: string;
    mimeType?: string;
}): Promise<DriveItemSummary[]> {
    const query: Record<string, unknown> = {};
    if (params?.folderId) query.folder_id = params.folderId;
    if (params?.mimeType) query.mime_type = params.mimeType;
    return request<DriveItemSummary[]>(
        "get",
        `${BASE}/tools/drive/files`,
        "Failed to list Drive files",
        query,
    );
}

export async function searchDriveFiles(
    query: string,
    mimeType?: string,
): Promise<DriveItemSummary[]> {
    const params: Record<string, unknown> = { query };
    if (mimeType) params.mime_type = mimeType;
    return request<DriveItemSummary[]>(
        "get",
        `${BASE}/tools/drive/search`,
        "Failed to search Drive",
        params,
    );
}

export async function listGmailLabels(): Promise<GmailLabelSummary[]> {
    return request<GmailLabelSummary[]>(
        "get",
        `${BASE}/tools/gmail/labels`,
        "Failed to list Gmail labels",
    );
}

export async function listGmailMessages(params?: {
    labelId?: string;
    maxResults?: number;
}): Promise<GmailMessageSummary[]> {
    const query: Record<string, unknown> = {};
    if (params?.labelId) query.label_id = params.labelId;
    if (params?.maxResults !== undefined) query.max_results = params.maxResults;
    return request<GmailMessageSummary[]>(
        "get",
        `${BASE}/tools/gmail/messages`,
        "Failed to list Gmail messages",
        query,
    );
}

export async function listGoogleDocs(): Promise<DocsSummary[]> {
    return request<DocsSummary[]>(
        "get",
        `${BASE}/tools/docs`,
        "Failed to list documents",
    );
}

File: api/schemas/google_workspace.py
"""Schemas for Google Workspace discovery — what a connected account contains.

These describe *pointers*, never content: a spreadsheet's id and title, a
calendar's name, a Drive file's link. Enough for an operator to pick the
resource an agent will be bound to, and nothing an untrusted conversation could
mine — the one exception is Gmail's ``snippet``, which Google returns as part of
the message list and which the UI shows to the account owner only.

Two conventions carry the whole contract:

- **A section is never just empty.** Every ``WorkspaceSectionResponse`` says
  *why* it has no items: ``scope_missing`` + ``missing_scopes`` when the
  organization's grant does not cover that API, ``error`` when Google failed,
  and neither when the account genuinely has nothing. A silent empty list is the
  bug this whole shape exists to prevent — the customer cannot tell "you never
  granted Calendar" from "you have no calendars".
- **Availability is per organization, not per deployment.** ``granted_scopes``
  is what *this* organization consented to; ``requested_scopes`` is what the
  deployment asks for at all. A scope in neither list cannot be fixed by
  re-consenting — the operator has to widen ``GOOGLE_OAUTH_SCOPES`` first (see
  ``services/integrations/google/scopes.py`` for what that costs).

Field names are snake_case, matching ``api/schemas/google_sheets.py`` and the
rest of the repo; the TypeScript client is generated from the OpenAPI these
models produce, so every field carries a description.
"""

from __future__ import annotations

from datetime import datetime
from typing import Dict, Generic, List, Optional, TypeVar

from pydantic import BaseModel, ConfigDict, Field

ItemT = TypeVar("ItemT")


# ---------------------------------------------------------------------------
# Resource summaries
# ---------------------------------------------------------------------------


class SpreadsheetSummary(BaseModel):
    """One Google Sheets file the account can reach."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(
        description=(
            "Drive file id, which is also the spreadsheetId the Sheets API takes."
        )
    )
    name: str = Field(description="Spreadsheet title as it reads in Drive.")
    web_url: Optional[str] = Field(
        default=None,
        description="Link that opens the spreadsheet in Google Sheets.",
    )
    modified_time: Optional[datetime] = Field(
        default=None, description="When the file was last modified."
    )


class SheetTabSummary(BaseModel):
    """One tab (worksheet) inside a spreadsheet."""

    model_config = ConfigDict(from_attributes=True)

    sheet_id: Optional[int] = Field(
        default=None,
        description=(
            "Numeric tab id assigned by Google. Stable across renames, unlike "
            "the title — but the title is what A1 ranges use."
        ),
    )
    title: str = Field(description="Tab name, e.g. 'Leads 2026'.")
    index: Optional[int] = Field(
        default=None, description="0-based position of the tab in the spreadsheet."
    )
    row_count: Optional[int] = Field(
        default=None, description="Grid height, including empty rows."
    )
    column_count: Optional[int] = Field(
        default=None, description="Grid width, including empty columns."
    )


class CalendarSummary(BaseModel):
    """One calendar in the account's calendar list."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(
        description=(
            "Calendar id used in Calendar API calls; 'primary' for the main one."
        )
    )
    summary: Optional[str] = Field(
        default=None, description="Human-readable calendar name."
    )
    description: Optional[str] = Field(
        default=None, description="Calendar description, when the owner set one."
    )
    primary: bool = Field(
        default=False, description="True for the account's primary calendar."
    )
    access_role: Optional[str] = Field(
        default=None,
        description=(
            "What the connected account may do here: 'owner', 'writer', "
            "'reader' or 'freeBusyReader'. Anything below 'writer' cannot host "
            "bookings."
        ),
    )
    color_id: Optional[str] = Field(
        default=None, description="Google's palette id, so the UI can match colours."
    )
    time_zone: Optional[str] = Field(
        default=None,
        description=(
            "IANA time zone of the calendar. Booking times are meaningless "
            "without it."
        ),
    )


class DriveItemSummary(BaseModel):
    """One Drive file or folder."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(description="Drive file or folder id.")
    name: str = Field(description="File or folder name.")
    mime_type: Optional[str] = Field(
        default=None,
        description=(
            "Google MIME type. 'application/vnd.google-apps.folder' marks a "
            "folder; see also `is_folder`."
        ),
    )
    is_folder: bool = Field(
        default=False,
        description=(
            "Convenience flag so the UI does not have to string-match the MIME "
            "type to decide whether an item can be opened."
        ),
    )
    web_url: Optional[str] = Field(
        default=None, description="Link that opens the item in Google Drive."
    )
    modified_time: Optional[datetime] = Field(
        default=None, description="When the item was last modified."
    )


class GmailLabelSummary(BaseModel):
    """One Gmail label."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(description="Label id used in Gmail API calls, e.g. 'INBOX'.")
    name: str = Field(description="Label name shown in Gmail.")
    type: Optional[str] = Field(
        default=None,
        description="'system' for Google's built-in labels, 'user' for custom ones.",
    )


class GmailMessageSummary(BaseModel):
    """Headers and snippet of one message — never its body."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(description="Message id.")
    thread_id: Optional[str] = Field(
        default=None, description="Thread the message belongs to."
    )
    subject: Optional[str] = Field(default=None, description="Subject header.")
    from_address: Optional[str] = Field(
        default=None,
        description=(
            "The From header, verbatim (display name and address). Named "
            "`from_address` because `from` is a Python keyword."
        ),
    )
    date: Optional[str] = Field(
        default=None,
        description=(
            "The Date header as Gmail returns it. Kept as text on purpose: "
            "senders emit malformed dates and a parse failure must not drop the "
            "message from the list."
        ),
    )
    snippet: Optional[str] = Field(
        default=None,
        description=(
            "Short preview Google includes in the message list. Shown to the "
            "account owner only."
        ),
    )


class DocsSummary(BaseModel):
    """One Google Docs document."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(description="Drive file id, which is the documentId for Docs.")
    name: str = Field(description="Document title.")
    web_url: Optional[str] = Field(
        default=None, description="Link that opens the document in Google Docs."
    )
    modified_time: Optional[datetime] = Field(
        default=None, description="When the document was last modified."
    )


# ---------------------------------------------------------------------------
# Workspace overview
# ---------------------------------------------------------------------------


class WorkspaceSectionFlags(BaseModel):
    """Why a section looks the way it does.

    Read the flags before the items: an empty section with ``scope_missing`` is
    a permission to ask for, an empty section with ``error`` is Google failing,
    and an empty section with neither is an account that has nothing there.
    """

    model_config = ConfigDict(from_attributes=True)

    available: bool = Field(
        default=False,
        description=(
            "True when this organization's grant covers the section's API and "
            "the listing succeeded. False means the items are empty for a "
            "reason given by the other fields."
        ),
    )
    scope_missing: bool = Field(
        default=False,
        description=(
            "True when the connected account never granted the scopes this "
            "section needs."
        ),
    )
    missing_scopes: List[str] = Field(
        default_factory=list,
        description=(
            "Exact scopes to add to the consent to make this section work. "
            "Empty when nothing is missing."
        ),
    )
    error: Optional[str] = Field(
        default=None,
        description=(
            "Set when the scopes were granted but the call to Google failed. "
            "One section failing never fails the whole overview."
        ),
    )
    notice: Optional[str] = Field(
        default=None,
        description=(
            "Set when the section worked but the grant narrows what it can "
            "see, so a short or empty list is expected rather than wrong. The "
            "case in practice is drive.file, which enumerates only the files "
            "this application was given — an account with a thousand "
            "spreadsheets legitimately shows none. Without this the flags say "
            "'available, no error, no items' and the owner concludes their "
            "files vanished."
        ),
    )


class WorkspaceSectionResponse(WorkspaceSectionFlags, Generic[ItemT]):
    """A section that is simply a list of one resource type."""

    items: List[ItemT] = Field(
        default_factory=list, description="Resources found, newest first."
    )


class WorkspaceDriveSection(WorkspaceSectionFlags):
    """Drive is shown as two lists: where to browse, and what changed lately."""

    recent_files: List[DriveItemSummary] = Field(
        default_factory=list,
        description="Most recently modified files, folders excluded.",
    )
    root_folders: List[DriveItemSummary] = Field(
        default_factory=list,
        description="Top-level folders of My Drive, as an entry point for browsing.",
    )


class WorkspaceGmailSection(WorkspaceSectionFlags):
    """Gmail is shown as its labels plus a short peek at the inbox."""

    labels: List[GmailLabelSummary] = Field(
        default_factory=list, description="Labels available for inbox automation."
    )
    recent: List[GmailMessageSummary] = Field(
        default_factory=list,
        description="A handful of the most recent inbox messages, headers only.",
    )


class WorkspaceOverviewResponse(BaseModel):
    """Everything a connected Google account exposes, in one call.

    Answers the question the UI asks right after the consent redirect — "what
    did I just connect, and what can it do?" — without five round-trips whose
    empty results cannot be told apart.
    """

    model_config = ConfigDict(from_attributes=True)

    connected: bool = Field(
        description=(
            "True when the organization has a usable, non-revoked Google "
            "connection. Every section is empty when this is false."
        )
    )
    google_email: Optional[str] = Field(
        default=None,
        description=(
            "Google account behind the connection, when known. Stays null when "
            "no identity scope was granted."
        ),
    )
    status: Optional[str] = Field(
        default=None,
        description=(
            "'connected' or 'revoked'; null when the organization never connected."
        ),
    )
    requires_reconsent: bool = Field(
        default=False,
        description=(
            "True when the grant is gone and the user must go through the "
            "consent screen again — refreshing cannot recover it."
        ),
    )
    granted_scopes: List[str] = Field(
        default_factory=list,
        description="Scopes this organization's consent actually covers.",
    )
    requested_scopes: List[str] = Field(
        default_factory=list,
        description=(
            "Scopes this deployment asks for at all (GOOGLE_OAUTH_SCOPES). A "
            "scope missing from here cannot be granted by re-consenting."
        ),
    )
    sheets: WorkspaceSectionResponse[SpreadsheetSummary] = Field(
        default_factory=WorkspaceSectionResponse[SpreadsheetSummary],
        description="Spreadsheets the account can reach.",
    )
    calendars: WorkspaceSectionResponse[CalendarSummary] = Field(
        default_factory=WorkspaceSectionResponse[CalendarSummary],
        description="Calendars in the account's calendar list.",
    )
    drive: WorkspaceDriveSection = Field(
        default_factory=WorkspaceDriveSection,
        description="Drive entry points: top-level folders and recent files.",
    )
    gmail: WorkspaceGmailSection = Field(
        default_factory=WorkspaceGmailSection,
        description="Gmail labels and a short inbox preview.",
    )
    docs: WorkspaceSectionResponse[DocsSummary] = Field(
        default_factory=WorkspaceSectionResponse[DocsSummary],
        description="Google Docs documents the account can reach.",
    )


# ---------------------------------------------------------------------------
# Scope registry
# ---------------------------------------------------------------------------


class GoogleScopeInfo(BaseModel):
    """One scope, with what it costs and who holds it."""

    model_config = ConfigDict(from_attributes=True)

    scope: str = Field(description="Full scope URL as Google spells it.")
    sensitivity: str = Field(
        description=(
            "Google's classification: 'non_sensitive' (nothing to do), "
            "'sensitive' (app verification), 'restricted' (verification plus a "
            "paid annual security assessment), or 'unknown' for a scope this "
            "deployment configured that the registry does not classify."
        )
    )
    requested: bool = Field(
        description="Whether this deployment's consent screen asks for the scope."
    )
    granted: bool = Field(
        default=False,
        description=(
            "Whether the calling organization's grant covers it, counting "
            "broader scopes that imply it."
        ),
    )


class GoogleRequiredScopesResponse(BaseModel):
    """The server's answer to 'which scopes does this tool need?'.

    The browser must not hardcode scope lists: which scopes a family needs is a
    server decision, and which ones this deployment may legally request is an
    operator decision. Both travel here.
    """

    model_config = ConfigDict(from_attributes=True)

    families: Dict[str, List[str]] = Field(
        description=(
            "Scopes per tool family — 'sheets', 'calendar', 'drive', 'gmail', "
            "'docs', plus 'all' for the union."
        )
    )
    requested_scopes: List[str] = Field(
        default_factory=list,
        description=(
            "What this deployment actually sends to Google's consent screen. "
            "Families listing a scope absent from here are not usable until an "
            "operator widens GOOGLE_OAUTH_SCOPES."
        ),
    )
    deployment_sensitivity: str = Field(
        default="non_sensitive",
        description=(
            "Heaviest classification among the requested scopes — what review "
            "this deployment owes Google as configured."
        ),
    )
    scope_details: List[GoogleScopeInfo] = Field(
        default_factory=list,
        description="Per-scope classification, so the UI can warn before widening.",
    )
